test/pkg/logging/logging.go

```diff
@@ -0,0 +1,169 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package logging
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"reflect"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"knative.dev/pkg/ptr"
+)
+
+type Logger interface {
+	Start() error
+}
+
+type logger struct {
+	labelsByNamespace map[string][]string
+	kc                kubernetes.Interface
+	ctx               context.Context
+}
+
+func NewLogger(ctx context.Context, client kubernetes.Interface, labelsByNamespace map[string][]string) Logger {
+	return &logger{
+		ctx:               ctx,
+		kc:                client,
+		labelsByNamespace: labelsByNamespace,
+	}
+}
+
+func (l *logger) Start() error {
+	if len(l.labelsByNamespace) == 0 {
+		return fmt.Errorf("you need to have at least one namespace to get logs from")
+	}
+
+	artifacts := os.Getenv("ARTIFACTS")
+
+	if _, err := os.Stat(fmt.Sprintf("%s/new-logs", artifacts)); errors.Is(err, os.ErrNotExist) {
+		err := os.Mkdir(fmt.Sprintf("%s/new-logs", artifacts), os.ModePerm)
+		if err != nil {
+			return err
+		}
+	}
+
+	for ns, ls := range l.labelsByNamespace {
+		requirement, err := labels.NewRequirement("app", selection.In, ls)
+		if err != nil {
+			return err
+		}
+
+		selector := labels.NewSelector().Add(*requirement)
+		watcher, err := l.kc.CoreV1().Pods(ns).Watch(l.ctx, metav1.ListOptions{LabelSelector: selector.String()})
+		if err != nil {
+			return err
+		}
+
+		go func() {
+			defer watcher.Stop()
+			watchedPods := sets.New[string]()
+			fileWriters := map[string]*os.File{}
+
+			for {
+				select {
+				case <-l.ctx.Done():
+					for _, writer := range fileWriters {
+						writer.Close()
+					}
+					return
+				case event := <-watcher.ResultChan():
+					if event.Object == nil || reflect.ValueOf(event.Object).IsNil() {
+						continue
+					}
+
+					pod, ok := event.Object.(*corev1.Pod)
+					if !ok {
+						continue
+					}
+
+					switch event.Type {
+					case watch.Deleted:
+						watchedPods.Delete(pod.Name)
+					case watch.Added, watch.Modified:
+						if !watchedPods.Has(pod.Name) && isPodReady(pod) {
+							label, ok := pod.Labels["app"]
+							if !ok {
+								continue
+							}
+							fileWriter, ok := fileWriters[label]
+							if !ok {
+								f, err := os.OpenFile(fmt.Sprintf("%s/new-logs/%s.log", artifacts, label), os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
```

Comment:
Can we create a separate file for each container in the pod? And also when a new pod starts, creating a new file (e.g. having the pod name in the filename)? I think this makes the usage a bit easier (IMO)